feat: Mapping component versions to RHOAI releases

Sara4994 opened the pull request:
`<!---` 
Many thanks for submitting your Pull Request ❤️!

Please complete the following sections for a smooth review.
-->

## Description
<!--- Describe your changes in detail -->

The PR is a first step to the feature initiative tracked in [JIRA](https://issues.redhat.com/browse/RHOAISTRAT-327) and this PR adds feature to capture RHOAI release versions from all the components(that are supported by DSC) repositories and include it in the DSC status. 

JIRA: https://issues.redhat.com/browse/RHOAIENG-12619
<!--- Link your JIRA and related links here for reference. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

To test with DevFlags:
- In a forked version of the components repository and create an .env file in config/manifest directory, have it in your branch.
- Point that forked repository as a DevFlag like [this](https://github.com/opendatahub-io/opendatahub-operator/wiki/4.-Using-Operator-in-Dev-Mode#devflags-example)
- Build and deploy the changes into the cluster.
- Create DSCI and DSC instances and you will be able see the status.components get updated with component upstream release details like below

`
apiVersion: [datasciencecluster.opendatahub.io/v1](http://datasciencecluster.opendatahub.io/v1)
kind: DataScienceCluster
metadata:

spec:
  components:
   
status:
 components:
  datasciencepipelines
   upstreamReleases:
    - name: datasciencepipelines
      displayName: Red Hat Build of Kubeflow Pipelines
      version: 2.2.0
      repoUrl: https://github.com/opendatahub-io/data-science-pipelines
 `

## Screenshot or short clip
<img width="985" alt="Screenshot 2024-10-26 at 12 12 59 PM" src="https://github.com/user-attachments/assets/81efaacd-8a36-41e5-85a2-d0e02dd6936c">




## Merge criteria
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

ykaliuta commented:
It's still draft, isn't it?

lburgazzoli commented on apis/datasciencecluster/v1/datasciencecluster_types.go:
```diff
@@ -114,7 +108,7 @@ type DataScienceClusterStatus struct {
 	// Expose component's specific status
 	// +optional
-	Components ComponentsStatus `json:"components,omitempty"`
+	Components map[string]status.ReleaseStatus `json:"components,omitempty"`
```
the `ComponentStatus` is meant to hold more information that the release status so I would not conflate different kind of information into a `ReleaseStaus` struct as it would be misleading. 

I would also recommend not to use a map as:
- the `ComponentStatus` should be symmetric to the `spec. Components` in term of naming 
- With a map, it would be extremely easy to accidental break the API

lburgazzoli commented on components/codeflare/codeflare.go:
```diff
@@ -72,6 +76,37 @@ func (c *CodeFlare) GetComponentName() string {
+		Version:     version.OperatorVersion{Version: componentVersion},
+		RepoURL:     repositoryURL}
+
+	fmt.Print("release object", componentReleaseStatus)
```
should be removed

lburgazzoli commented on components/component.go:
```diff
@@ -86,6 +87,7 @@ type ComponentInterface interface {
 		owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec, platform cluster.Platform, currentComponentStatus bool) error
 	Cleanup(ctx context.Context, cli client.Client, owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec) error
 	GetComponentName() string
+	GetComponentStatus() ([]status.ComponentReleaseStatus, error)
```
It may be better to pass a `DataScienceCluster` instance to the method, so each component can set the related status according to the specif need.

lburgazzoli replied:
Thinking a little bit more, it may even be moved in the `ReconcileComponent` method (with some method signature changes)

Sara4994 replied:
@lburgazzoli this approach didnt work as expected. I am unable to pass *dscv1.DataScienceCluster to ReconcileComponent method. getting error `import cycle not allowedgo list
Package v1 contains API Schema definitions for the datasciencecluster v1 API group`

lburgazzoli commented:
@Sara4994 I'd like to see this kind of steps to be performed only once i.e. at initialization time as component version won't change during the lifetime of the operator.

ykaliuta commented on components/codeflare/codeflare.go:
```diff
@@ -71,6 +75,35 @@ func (c *CodeFlare) GetComponentName() string {
 	return ComponentName
 }
 
+func (c *CodeFlare) GetUpstreamReleaseStatus() status.UpstreamReleases {
```
Please, redo it to avoid such a code duplication

ykaliuta commented on components/datasciencepipelines/datasciencepipelines.go:
```diff
@@ -92,6 +95,35 @@ func (d *DataSciencePipelines) GetComponentName() string {
+	var releaseDetails = make([]status.ComponentReleaseStatus, 0)
+	var releaseStatus status.UpstreamReleases
+
+	env, err := godotenv.Read(filepath.Join(deploy.DefaultManifestPath, ComponentName, ".env"))
```
We have similar functionality related to ApplyParams().Do we need to bring another dependency?

zdtsw replied:
i am not fond of introducing godotenv. but does not feel like whats in ApplyParams can be easiliy work as "read 
 file+convert to map"

zdtsw replied:
i leave this to @ykaliuta for more comments

ykaliuta replied:
> i am not fond of introducing godotenv. but does not feel like whats in ApplyParams can be easiliy work as "read file+convert to map"

doesn't parseParams() do exactly that?

ykaliuta replied:
So, the file as it is now actually in a custom format with spaces around `=` and `"` around values. Not what I would expect from `env`, but go has own opinion :)

zdtsw replied:
i am on the side of using key=value, not to have wider support for both key=value and key:value
which i think godotenv supporr the later case.
but since i do not see any detail in the refinement doc, except we need env file, so i do not know what has been agreed with all stakeholders, thats why i think the parseParams() only work with "="  and if " is needed.

Sara4994 replied:
@ykaliuta even with strict format (KEY=VALUE), it get this error with parseParams() `Invalid character(s) found in major number "\"1"`

ykaliuta replied:
> @ykaliuta even with strict format (KEY=VALUE), it get this error with parseParams() `Invalid character(s) found in major number "\"1"`

probably, you still use `"` which in _strict_ format becomes part of the value.

Sara4994 replied:
@ykaliuta yes my bad, after removing quotes now it is working

ykaliuta commented on components/ray/ray.go:
```diff
@@ -68,6 +73,35 @@ func (r *Ray) GetComponentName() string {
+		return status.UpstreamReleases{}
+	}
+	if env != nil {
+		componentVersion, err = semver.Parse(env["RHOAI_RELEASE_VERSION"])
```
Why can't this variable names be the same across all the components? Only ray and dashboard special for some reason

zdtsw commented on components/codeflare/codeflare.go:
```diff
@@ -71,9 +72,32 @@ func (c *CodeFlare) GetComponentName() string {
+	in.CodeFlare = &status.CodeFlareStatus{
+		ComponentStatus: status.ComponentStatus{
+			UpstreamReleases: []status.ComponentReleaseStatus{{
+				Name:        status.Platform(ComponentName),
+				DisplayName: CodeflareOperator,
```
any reason to keep both?
or maybe share the vision after this 1st phase, what will be built evenetually so we do not need to deprcated fields later once we add here

Sara4994 replied:
@zdtsw this was discussed in the refinement [here](https://docs.google.com/document/d/1nbQB-uA48x79Ci3xrpMHGfdo3XjTR4xirtWk0we0kl8/edit?disco=AAABP_8ne1g)

zdtsw commented on components/component.go:
```diff
@@ -195,3 +204,23 @@ func (c *Component) UpdatePrometheusConfig(_ client.Client, logger logr.Logger,
+		return ComponentReleaseDetails{}, err
+	}
+
+	componentVersion, err = semver.Parse(env["RHOAI_RELEASE_VERSION"])
```
what if it is not set in the file, componentVersion is going to be 0.0.0 ?

zdtsw commented on components/codeflare/codeflare.go:
```diff
@@ -71,9 +72,32 @@ func (c *CodeFlare) GetComponentName() string {
+	codeFlareStatus, err := c.GetReleaseVersion(in, deploy.DefaultManifestPath, ComponentName)
+
+	if err != nil {
+		in.CodeFlare = &status.CodeFlareStatus{}
```
1. in dsc controller the error is not checked, as it throw err or not does not really matter?
2. in the error case, it still is going to set the status for component ? 
```
Name : ""
DsiplayName: ""
Version: 0.0.0
RepoURL: ""
```

Sara4994 replied:
in case of error or no upstream, just returns empty struct and displays like dashboard: {} and it is handled inside the component itself. So i feel dsc controller doesnt have to handle the error.

zdtsw commented on components/component.go:
```diff
@@ -80,12 +83,18 @@ type ManifestsConfig struct {
 	Init(ctx context.Context, platform cluster.Platform) error
-	ReconcileComponent(ctx context.Context, cli client.Client,
-		owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec, platform cluster.Platform, currentComponentStatus bool) error
+	ReconcileComponent(ctx context.Context, cli client.Client, owner client.Object, DSCISpec *dsciv1.DSCInitializationSpec,
```
any reason to change to client.Object for this PR?

lburgazzoli replied:
it was probably a consequence of some discussions about moving the logic in the `ReconcileComponent` but I think it should now be reverted.

VaishnaviHire commented on components/datasciencepipelines/datasciencepipelines.go:
```diff
@@ -92,9 +92,32 @@ func (d *DataSciencePipelines) GetComponentName() string {
+		ComponentStatus: status.ComponentStatus{
+			UpstreamReleases: []status.ComponentReleaseStatus{{
+				Name:        status.Platform(ComponentName),
+				DisplayName: ComponentName,
```
Should the display name be something that is provided by the component? I believe upstream component name would be different than component name?

Sara4994 commented on components/component.go:
```diff
@@ -195,3 +204,23 @@ func (c *Component) UpdatePrometheusConfig(_ client.Client, logger logr.Logger,
+	var componentVersion semver.Version
+	var repositoryURL string
+
+	env, err := godotenv.Read(filepath.Join(defaultManifestPath, componentName, ".env"))
```
yes correct

zdtsw commented:
> @Sara4994 I'd like to see this kind of steps to be performed only once i.e. at initialization time as component version won't change during the lifetime of the operator.

has this been addressed?
if i understand the current code, the `read-get-set` action is still done per every enabled successfully reconciled components

zdtsw commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -335,8 +335,11 @@ func (r *DataScienceClusterReconciler) reconcileSubComponent(ctx context.Context
+
 	// reconciliation succeeded: update status accordingly
 	instance, err = status.UpdateWithRetry(ctx, r.Client, instance, func(saved *dscv1.DataScienceCluster) {
+		_ = component.UpdateStatus(&saved.Status.Components)
```
after the internal discussion, we will only set the compoent version if the component is enabled.

ykaliuta commented on components/codeflare/codeflare.go:
```diff
@@ -71,6 +73,29 @@ func (c *CodeFlare) GetComponentName() string {
+	}
+
+	in.CodeFlare = &status.CodeFlareStatus{
+		ComponentStatus: status.ComponentStatus{
```
would it be possible to move whole creation of ComponentStatus in one separate function leaving differences to parameters and methods?

Sara4994 replied:
@ykaliuta this is done

ykaliuta commented on controllers/status/status.go:
```diff
@@ -210,7 +212,95 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
 }
 
-// ModelRegistryStatus struct holds the status for the ModelRegistry component.
+type Platform string
```
The conversation is resolved but sorry, I missed the answer. Having component short name of type "Platform" sounds confusing at the first glance.

ykaliuta commented on Dockerfiles/Dockerfile:
```diff
@@ -39,7 +39,7 @@ COPY main.go main.go
 
 # Build
-RUN CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
```
Nope

Sara4994 commented:
> > @Sara4994 I'd like to see this kind of steps to be performed only once i.e. at initialization time as component version won't change during the lifetime of the operator.
> 
> has this been addressed? if i understand the current code, the `read-get-set` action is still done per every enabled successfully reconciled components

@zdtsw @lburgazzoli i tried few things to make this 'read-get-set' logic to run once during component initialization time. but it doesnt work as expected with devFlags. Because devflags gets the updates the reconcile time and so i get error saying there is no env file `/opt/manifests/codeflare/.env: no such file or directory`

ykaliuta commented on pkg/deploy/envParams.go:
```diff
@@ -121,3 +127,34 @@ func ApplyParams(componentPath string, imageParamsMap map[string]string, extraPa
+	var repositoryURL string
+	var displayName string
+
+	env, err := parseParams(filepath.Join(defaultManifestPath, componentName, ".env"))
```
probably should be moved to some common place and exported. Putting GetReleaseVersion() probably not the best.

ykaliuta commented on pkg/deploy/envParams.go:
```diff
@@ -121,3 +127,34 @@ func ApplyParams(componentPath string, imageParamsMap map[string]string, extraPa
+
+	return status.ComponentStatus{
+		UpstreamReleases: []status.ComponentReleaseStatus{{
+			Name:        cluster.Platform(componentName),
```
still did not get type cluster.Platform for Name

ykaliuta commented on go.sum:
```diff
@@ -219,6 +219,8 @@ github.com/ianlancetaylor/demangle v0.0.0-20181102032728-5e5cf60278f6/go.mod h1:
 github.com/imdario/mergo v0.3.13 h1:lFzP57bqS/wsqKssCGmtLAb8A0wKjLGrve2q3PPVcBk=
 github.com/imdario/mergo v0.3.13/go.mod h1:4lJ1jqUDcsbIECGy0RUJAXNIhg+6ocWgb1ALK2O4oXg=
+github.com/joho/godotenv v1.5.1 h1:7eLL/+HRGLY0ldzfGMeQkb7vMd0as4CfYvUVzLqw0N0=
+github.com/joho/godotenv v1.5.1/go.mod h1:f4LDr5Voq0i2e/R5DDNOoa2zzDfwtkZa6DnEwAbqwq4=
```
still?

ykaliuta commented on components/codeflare/codeflare.go:
```diff
@@ -71,6 +72,21 @@ func (c *CodeFlare) GetComponentName() string {
+	}
+
+	in.CodeFlare = &status.CodeFlareStatus{
+		ComponentStatus: codeFlareStatus,
```
will unconditional setting this with return `err` and avoiding the first check make any difference?

ykaliuta commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -335,13 +335,15 @@ func (r *DataScienceClusterReconciler) reconcileSubComponent(ctx context.Context
 		})
 		return instance, err
 	}
+
```
Unrelated change

ykaliuta commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -335,13 +335,15 @@ func (r *DataScienceClusterReconciler) reconcileSubComponent(ctx context.Context
 		}
 		saved.Status.InstalledComponents[componentName] = enabled
 		if enabled {
+			_ = component.UpdateStatus(&saved.Status.Components)
```
so, why do you return error at all if it's ignored by the only user?

ykaliuta commented on components/component.go:
```diff
@@ -82,10 +86,11 @@ type ManifestsConfig struct {
 	Init(ctx context.Context, platform cluster.Platform) error
-	ReconcileComponent(ctx context.Context, cli client.Client,
-		owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec, platform cluster.Platform, currentComponentStatus bool) error
+	ReconcileComponent(ctx context.Context, cli client.Client, owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec,
```
not sure what the change is about. Reformatting? Why?

ykaliuta replied:
I will not block the PR, but the questions are marked resolved without reverting the unrelated changes nor explanation.

Sara4994 replied:
Sorry that wasn't intentional, it was auto-formatted for some reason. Now it is reverted. thanks

ykaliuta commented on components/component.go:
```diff
@@ -195,3 +200,34 @@ func (c *Component) UpdatePrometheusConfig(_ client.Client, logger logr.Logger,
 	return err
 }
+
+func (c *Component) GetReleaseVersion(defaultManifestPath string, componentName string) status.ComponentStatus {
```
Frankly, I do not see a point to have it as a method in such form. It's just a utility function. It probably could if it used c->GetComponentName() instead of the componentName argument, I'm not sure if it fits all cases.

(but fine for me if others ok)

ykaliuta commented on components/dashboard/dashboard.go:
```diff
@@ -82,6 +83,10 @@ func (d *Dashboard) GetComponentName() string {
 }
 
+func (d *Dashboard) UpdateStatus(in *status.ComponentsStatus) {
+	in.Dashboard = &status.DashboardStatus{}
```
Did not check too deep, why dashboard is special?

Sara4994 replied:
Dashboard dont have an upstream for now. so keeping it empty for now

ykaliuta commented on controllers/status/status.go:
```diff
@@ -210,7 +212,93 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
 }
 
-// ModelRegistryStatus struct holds the status for the ModelRegistry component.
+type ComponentObject interface {
```
Is it used anywhere? grep does not show any place.

Sara4994 replied:
removed

ykaliuta commented on controllers/status/status.go:
```diff
@@ -210,7 +212,93 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+
+// +k8s:deepcopy-gen=true
+type ComponentStatus struct {
+	UpstreamReleases []ComponentReleaseStatus `json:"upstreamrelease,omitempty"`
```
I probably missed something, could you say, why it is and array?

Should it be "Upstream"? Looks like just a release of the component.

Sara4994 replied:
@ykaliuta Well, it was discussed that there could be multiple upstreams for the components in future. So made it as an array to accommodate those

zdtsw replied:
any reason we want to name it as `upstreamrelease`?
i remember it was called release or something days ago.
since we already have display and ulr under this filed, with the name upstream can cause more confusion.
e.g 'https://github.com/opendatahub-io/kuberay' is not the upstream for kuberay, rather a midstream for rhoai component

Sara4994 replied:
@zdtsw yea makes sense I am ok to change it to "releases" if everyone agrees on it  @VaishnaviHire

ykaliuta replied:
> @zdtsw yea makes sense I am ok to change it to "releases" if everyone agrees on it @VaishnaviHire

I asked about it above :)

Sara4994 replied:
lol yea wrong tag.

Sara4994 replied:
@zdtsw @ykaliuta this is done

ykaliuta commented on apis/datasciencecluster/v1/datasciencecluster_types.go:
```diff
@@ -167,3 +161,7 @@ func (d *DataScienceCluster) GetComponents() ([]components.ComponentInterface, e
 	return allComponents, nil
 }
+
+func (d *DataScienceCluster) GetComponentsStatus() *status.ComponentsStatus {
```
Is it used anywhere? grep does not show any.

ykaliuta commented on components/component.go:
```diff
@@ -195,3 +201,34 @@ func (c *Component) UpdatePrometheusConfig(_ client.Client, logger logr.Logger,
 	return err
 }
+
+func GetReleaseVersion(defaultManifestPath string, componentName string) status.ComponentStatus {
```
I will not block the PR, but with that the functions looks like `GetComponentStatus` and belongs status package.

Sara4994 replied:
this is done

VaishnaviHire commented:
/test all

VaishnaviHire commented:
Hi @Sara4994 , when I update the component from `Managed` --> `Removed`, it still shows upstream versions in status. E.g codeflare


```
spec:
  components:
    codeflare:
      devFlags:
        manifests:
          - contextDir: config
            sourcePath: ''
            uri: 'https://github.com/sara4994/codeflare-operator/tarball/codeflare-release-version'
      managementState: Removed
    kserve:
      managementState: Removed
      serving:
        ingressGateway:
          certificate:
            type: SelfSigned
        managementState: Removed
        name: knative-serving
    modelregistry:
      managementState: Removed
      registriesNamespace: odh-model-registries
    trustyai:
      managementState: Removed
    ray:
      devFlags:
        manifests:
          - contextDir: ray-operator/config
            sourcePath: ''
            uri: 'https://github.com/sara4994/kuberay/tarball/ray-release-version'
```

The status shown is as follows
```
status:
  components:
    codeflare:
      upstreamrelease:
        - name: codeflare
          version: 1.0.0
    dashboard: {}
    modelregistry: {}
    ray:
      upstreamrelease:
        - displayname: KubeRay
          name: ray
          repourl: 'https://github.com/opendatahub-io/kuberay'
          version: 1.0.0
```

Sara4994 commented:
> Hi @Sara4994 , when I update the component from `Managed` --> `Removed`, it still shows upstream versions in status. E.g codeflare
> 
> ```
> spec:
>   components:
>     codeflare:
>       devFlags:
>         manifests:
>           - contextDir: config
>             sourcePath: ''
>             uri: 'https://github.com/sara4994/codeflare-operator/tarball/codeflare-release-version'
>       managementState: Removed
>     kserve:
>       managementState: Removed
>       serving:
>         ingressGateway:
>           certificate:
>             type: SelfSigned
>         managementState: Removed
>         name: knative-serving
>     modelregistry:
>       managementState: Removed
>       registriesNamespace: odh-model-registries
>     trustyai:
>       managementState: Removed
>     ray:
>       devFlags:
>         manifests:
>           - contextDir: ray-operator/config
>             sourcePath: ''
>             uri: 'https://github.com/sara4994/kuberay/tarball/ray-release-version'
> ```
> 
> The status shown is as follows
> 
> ```
> status:
>   components:
>     codeflare:
>       upstreamrelease:
>         - name: codeflare
>           version: 1.0.0
>     dashboard: {}
>     modelregistry: {}
>     ray:
>       upstreamrelease:
>         - displayname: KubeRay
>           name: ray
>           repourl: 'https://github.com/opendatahub-io/kuberay'
>           version: 1.0.0
> ```

@VaishnaviHire Thanks for pointing it. this is fixed now :)

VaishnaviHire commented:
/lgtm

Will defer to other reviewers for approval

zdtsw commented on controllers/status/status.go:
```diff
@@ -210,7 +216,119 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+	TrainingOperator     *TrainingOperatorStatus     `json:"trainingoperator,omitempty"`
+	Workbenches          *WorkbenchesStatus          `json:"workbenches,omitempty"`
+}
+
```
```suggestion
// GetReleaseVersion read .env file and parse env variables delimiter by "=". If version is not set or set to "", return empty {}
```

Sara4994 replied:
added this line

zdtsw commented on controllers/status/status.go:
```diff
@@ -210,7 +216,119 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+	Workbenches          *WorkbenchesStatus          `json:"workbenches,omitempty"`
+}
+
+func GetReleaseVersion(defaultManifestPath string, componentName string) ComponentStatus {
```
why not move this function into pkg/cluster/cluster_config.go

Sara4994 replied:
@zdtsw Earlier i had it in components.go since it is an utility share by component. later on suggestions i moved it to controllers/status/status.go and it looks bit more appropriate, as this is used my components to display releases in status. wdyt?

zdtsw replied:
no strong opinion to keep it here or not.

zdtsw commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -349,7 +350,7 @@ func (r *DataScienceClusterReconciler) reconcileSubComponent(ctx context.Context
 			if enabled {
 				saved.Status.Components.ModelRegistry = &status.ModelRegistryStatus{RegistriesNamespace: mr.RegistriesNamespace}
 			} else {
-				saved.Status.Components.ModelRegistry = nil
+				saved.Status.Components.ModelRegistry = &status.ModelRegistryStatus{}
 			}
```
why not move this into f`unc (m *ModelRegistry) UpdateStatus(in *status.ComponentsStatus`) ?

Sara4994 replied:
yea nice catch, just moved it thanks :)

zdtsw commented:
do not checkin `manager` binary

zdtsw reviewed (commented):
some small parts, otherwise i am fine with the changes.

Sara4994 commented on controllers/status/status.go:
```diff
@@ -210,7 +216,119 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+		return ComponentStatus{}
+	}
+
+	componentVersion, err = semver.Parse(env["RHOAI_RELEASE_VERSION"])
```
sure, i will check with teams or may be we can leave it to them to decide the name(to name with specific upstream names) and here we can the pass it as an arg.

zdtsw replied:
you've lost me.
if the name has not been decided with stakeholders, how can we put RHOAI_RELEASE_VERSION here?

Sara4994 replied:
well. Earlier it was discussed to provide naming suggestions(envs) from our end. but still i can go back to the teams to gather their inputs, if that would be better.

Sara4994 commented:
@ykaliuta @zdtsw @VaishnaviHire @lburgazzoli I have tweaked `GetReleaseVersion` function to read release info from an yaml file instead of an env. 

Reason: few components have got multiple upstreams, reading release info of multiple upstreams from an env would be tricky, whereas with yaml we could group the info based on the upstream, and it would be easy to fetch the details at our end. Please feel free to share your thoughts.

Sara4994 commented:
/cc @grdryn

grdryn commented on controllers/status/status.go:
```diff
@@ -210,7 +216,135 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
-// ModelRegistryStatus struct holds the status for the ModelRegistry component.
+// +k8s:deepcopy-gen=true
+type ComponentReleaseStatus struct {
+	DisplayName string                  `json:"displayname,omitempty"`
+	Version     version.OperatorVersion `json:"version,omitempty"`
+	RepoURL     string                  `json:"repourl,omitempty"`
```
The JSON keys here could maybe be slightly improved (it's probably just me, but I had a tendancy to read `repourl` as "repour L" rather than "repo URL" :sweat_smile:)

```suggestion
	DisplayName string                  `json:"displayName,omitempty"`
	Version     version.OperatorVersion `json:"version,omitempty"`
	RepoURL     string                  `json:"repoURL,omitempty"`
```

grdryn replied:
Related to this, your screenshot in the description also has a "name" field, where does that come from, or is it no longer included? If it's no longer included, then maybe displayName here should just become name?

Sara4994 replied:
here the field "name" would contain component name, which is already present a level above(like below) so we wouldnt need have "name" field again. I will update the screenshot now.

```
status:
   components:
        codeflare:
             - displayname: 
                version:
                repoURL:
```

grdryn replied:
That's the name of the component from the ODH operator POV, but the name/displayname for each element in the list is the name/displayname for the "upstream release project" (or whatever), right? So I'd suggest changing the "displayname" field of the elements in the "releases" list here to just be "name" if there's just going to be one key to represent this.

Sara4994 replied:
yea makes sense. will change it to "name"

grdryn reviewed (commented):
/hold

(at least until ./manager is removed from the patch set)

Sara4994 commented on pkg/common/common.go:
```diff
@@ -116,3 +117,26 @@ func GetMonitoringData(data string) (string, error) {
 	return encodedData, nil
 }
+
+func ParseParams(fileName string) (map[string]string, error) {
```
thats a good catch. just restored it back.

Sara4994 commented on components/codeflare/codeflare.go:
```diff
@@ -71,6 +72,19 @@ func (c *CodeFlare) GetComponentName() string {
 
+func (c *CodeFlare) UpdateStatus(in *status.ComponentsStatus) {
+	enabled := c.GetManagementState() == operatorv1.Managed
+	if enabled {
+		codeFlareStatus := status.GetReleaseVersion(deploy.DefaultManifestPath, ComponentName)
+
+		in.CodeFlare = &status.CodeFlareStatus{
+			ComponentStatus: codeFlareStatus,
+		}
+	} else {
+		in.CodeFlare = nil
+	}
```
but if we do this way, even if the "enabled" is true first it will set "nil"(always) and changes it to codeFlareStatus.

grdryn replied:
Are you saying that there's there a potential value there that you care about (I don't think so), or that there's a performance hit to assigning to `nil`?

In any case, I don't think this is an important thing either way, so I'll mark this as resolved :+1:

Sara4994 commented on controllers/status/status.go:
```diff
@@ -210,7 +216,135 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+	}
+
+	for _, release := range releaseInfo.Releases {
+		componentVersion, err = semver.Parse(release.Version)
+
+		if err != nil {
+			return ComponentStatus{}
+		}
```
I see other versions(like [operator version](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/pkg/cluster/cluster_config.go#L197)) maintained in repo are of type  "semver.Version". so thought to have these component versions in same type as well. This parse method would parse the string to proper version format(major.minor.patch)

grdryn replied:
I don't have strong opinions on it, so I'll leave it to others to decide on it one way or another.

Here we're just reading in a string value and outputting a string value, and I don't think we ever need to use the individual parts of the version string, or care if they follow the semver approach (which maybe they always will, but I don't know if we have any control over that).

Sara4994 commented on manager:
this is removed now

grdryn commented:
/unhold

It looks like, for better or worse, the merge automation in this repo will squash all commits in a PR into a single commit, when merging, so the fact that the manager binary was added and then removed in separate commits won't matter: those should cancel each other out and it won't exist in the final history.

grdryn commented on controllers/status/status.go:
```diff
@@ -210,7 +216,135 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+type ComponentReleaseStatusMeta struct {
+	name    string `yaml:"name,omitempty"`
+	Version string `yaml:"version,omitempty"`
+	RepoURL string `yaml:"repourl,omitempty"`
```
```suggestion
	RepoURL string `yaml:"repoURL,omitempty"`
```

at least in the ComponentReleaseStatus on L223, it was changed like this. Could alternatively change to `repository` if the capitalization is confusing.

grdryn commented on controllers/status/status.go:
```diff
@@ -210,7 +216,135 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+
+// +k8s:deepcopy-gen=true
+type ComponentReleaseStatusMeta struct {
+	name    string `yaml:"name,omitempty"`
```
```suggestion
	Name    string `yaml:"name,omitempty"`
```

I guess?

lburgazzoli commented on controllers/status/status.go:
```diff
@@ -210,7 +216,135 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+
+// GetReleaseVersion reads odh_metadata.yaml file and parses release information.
+// If version is not set or set to "", return empty {}.
+func GetReleaseVersion(defaultManifestPath string, componentName string) ComponentStatus {
```
I think `GetReleaseVersion` should return a `[]ComponentReleaseStatus` not the entire `ComponentStatus`

Sara4994 replied:
@lburgazzoli but incase of yaml file not present in the components repo, we discussed to show empty struct against component names. such error conditions are handled in the functions itself so as returning empty struct. When returning `[]ComponentReleaseStatus` , will have to handle error cases in components.

lburgazzoli replied:
you can return an empty list, if you sue `omitempty` in the tags, then it won't not show up

Sara4994 replied:
@lburgazzoli yep made this changes now. :)

ykaliuta replied:
> I think `GetReleaseVersion` should return a `[]ComponentReleaseStatus` not the entire `ComponentStatus`

If it returns ReleaseStatus, why it's still called GetReleaseVersion? (asked @Sara4994 already)

Sara4994 replied:
@ykaliuta renamed it now. thanks

lburgazzoli commented on controllers/status/status.go:
```diff
@@ -210,7 +216,135 @@ func RemoveComponentCondition(conditions *[]conditionsv1.Condition, component st
+	componentReleaseStatus := make([]ComponentReleaseStatus, 0)
+
+	yamlData, err := os.ReadFile(filepath.Join(defaultManifestPath, componentName, "odh_metadata.yaml"))
+	if err != nil {
```
I think we need to distinguish between a file not present from an error, in this case we are swallowing it and I'm not sure that's the right approach as it could hide some other errors.

Sara4994 commented:
opened a new [PR](https://github.com/opendatahub-io/opendatahub-operator/pull/1486) for this, applying code changes on top of refactored components. Hence closing this one.
